src/v/net/conn_quota.cc

```diff
@@ -0,0 +1,440 @@
+    bool was_dirty_decrease = in_use != 0 && allowance.max > new_limit;
+
+    allowance.max = new_limit;
+    if (in_use >= allowance.max) {
+        vlog(
+          rpc::rpclog.trace,
+          "Connection count limit {} decreased below current ({}) for {}",
+          allowance.max,
+          in_use,
+          addr);
+        allowance.available = 0;
+    } else {
+        allowance.available = allowance.max - in_use;
+    }
+
+    // If the allowance might have had borrowed units on other
+    // nodes, then we must reclaim them in case of a decrease
+    // to the limit.
+    if (was_dirty_decrease) {
+        ssx::spawn_with_gate(_gate, [this, addr]() {
+            auto& allowance = get_home_allowance(addr);
+            return reclaim_to(allowance, addr, true);
+        });
+    }
+}
+
+conn_quota::home_allowance&
+conn_quota::get_home_allowance(ss::net::inet_address addr) {
+    assert_on_home(addr);
+
+    if (addr == ss::net::inet_address{}) {
+        return total_home;
+    } else {
+        auto found = ip_home.find(addr);
+        if (found != ip_home.end()) {
+            return found->second;
+        } else {
+            auto [iter, created] = ip_home.insert(std::make_pair(
+              addr,
+              home_allowance{
+                .max = _cfg.max_connections_per_ip().value(),
+                .available = _cfg.max_connections_per_ip().value()}));
+            return iter->second;
+        }
+    }
+}
+
+conn_quota::remote_allowance&
+conn_quota::get_remote_allowance(ss::net::inet_address addr) {
+    if (addr == ss::net::inet_address{}) {
+        return total_remote;
+    } else {
+        auto found = ip_remote.find(addr);
+        if (found != ip_remote.end()) {
+            return found->second;
+        } else {
+            auto [iter, created] = ip_remote.insert(
+              std::make_pair(addr, remote_allowance{}));
+            return iter->second;
+        }
+    }
+}
+
+/**
+ * Called on the home shard for `addr`.  Dispatch reclaim requests
+ * to all other shards, hopefully they had some borrowed tokens
+ * that we can claw back.
+ */
+ss::future<> conn_quota::reclaim_to(
+  conn_quota::home_allowance& allowance,
+  ss::net::inet_address addr,
+  bool one_time) {
+    // Note: require that allowance is pointer-stable for this lock
+    // to be safely freed later (i.e. the ip_home container must
+    // provide pointer stability)
+
+    auto locked = allowance.reclaim_lock.get_units();
+    if (allowance.reclaim) {
+        // We are already in reclaim mode: remote allowances will
+        // not be holding any units belong to us, so don't waste
+        // time looking.
+        co_return;
+    }
+
+    allowance.reclaim = true;
+    uint32_t total_released = 0;
+    for (ss::shard_id i = 0; i < ss::smp::count; ++i) {
+        if (i == ss::this_shard_id()) {
+            continue;
+        }
+
+        total_released += co_await container().invoke_on(
```

Comment:
I think this is the same as `container().invoke_on_others()`

Reply:
Is that safe?  I think if I passed the `total_released` reference into invoke_on_others then it would have to be an atomic.

Reply:
Sorry, I wasn't being precise. I think it is the same as whatever the `invoke_on_others` variant is for `ss::sharded<T>::map_reduce0` but looking now that variant doesn't seem to exist :( So the best I think we could do without building a new utility would be:

```
total_released = co_await container().map_reduce0(
   [home = ss::this_shard_id()](...) { if (home == ss::this_hard_id()) { return 0; } ... },
   0,
   std::add<>);
```

or something like that. doesn't seem necessary to change tho.

Reply:
I've gone ahead and changed this to use a map_reduce0 to improve the latency of the reclaim, it was a bit lazy to serialize those calls out to other cores.